### PR TITLE
CAMS-526 no library extended

### PR DESCRIPTION
**Pull request by fmaddenflx**
# Purpose

Add Rich Text Formatting to Case Notes

# Major Changes

New RichTextEditor.tsx, Editor.ts, SelectionService humble object for document.range and window.selection, RichTextButton and RichTextIcon.

# Testing/Validation

Test coverage is above 90%

# Notes

This includes a lot of low level DOM manipulation working with content editable div and a very high usage of AI.

# Definition of Done:

- [ ] Code refactored for clarity: Developers can understand the work simply by reviewing the code
- [ ] Dependency rule followed: More important code doesn’t directly depend on less important code
- [ ] Development debt eliminated: UX and code aligns to the team’s latest understanding of the domain
- [ ] No regressions: Changes do not cause regression in related or unrelated areas of the application

## Summary by Sourcery

Integrate a custom rich text editor into case notes, replacing the existing textarea with a contenteditable component that supports bold, italic, underline, and list formatting, and adapt form logic and tests accordingly.

New Features:
- Introduce RichTextEditor component with toolbar buttons for bold, italic, underline, bulleted list, and numbered list formatting
- Replace plain textarea in CaseNoteFormModal with the rich text editor and submit HTML-wrapped content

Enhancements:
- Abstract selection handling via BrowserSelectionService and MockSelectionService for consistent DOM manipulation
- Sanitize and clean editor HTML output using DOMPurify and utility methods in Editor
- Refine Input and ButtonGroup components to properly handle className propagation and required attributes
- Rename input value helpers in modal to clearly separate title and content getters

Tests:
- Add a comprehensive Vitest suite for Editor covering initialization, formatting toggles, list indentation/outdentation, key events, and edge cases
- Update CaseNoteFormModal tests to use contenteditable selectors and assert paragraph-wrapped HTML content

### CODE REVIEW

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  const setValue = (html: string) => {
+    if (contentRef.current) {
+      if (html.trim() === '') {
+        // If setting empty content, reinitialize with empty paragraph
+        contentRef.current.innerHTML = '';
+        if (editorRef.current) {
+          editorRef.current = new Editor(contentRef.current, selectionService);
+        }
+      } else {
+        contentRef.current.innerHTML = html;
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx]: Please reevaluate based on new code changes.

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  const setValue = (html: string) => {
+    if (contentRef.current) {
+      if (html.trim() === '') {
+        // If setting empty content, reinitialize with empty paragraph
+        contentRef.current.innerHTML = '';
+        if (editorRef.current) {
+          editorRef.current = new Editor(contentRef.current, selectionService);
+        }
+      } else {
+        contentRef.current.innerHTML = DOMPurify.sanitize(html);
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:83]: we are sanitizing the user controlled data with DOMPurify.

File: user-interface/src/lib/components/cams/RichTextEditor/editor.constants.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export const ZERO_WIDTH_SPACE = '​';
+export const ZERO_WIDTH_SPACE_REGEX = new RegExp(ZERO_WIDTH_SPACE, 'g');
+export const EMPTY_TAG_REGEX = /<([a-z][^>]*)>\s*(?:<br\s*\/?>)*\s*<\/\1>/gi;
+
+export type RichTextFormat = 'strong' | 'em' | 'u';
```
Comment (En-8) [inline user-interface/src/lib/components/cams/RichTextEditor/editor.constants.ts:5]: We should double-check that each of these constants are actually used and not redefined elsewhere unnecessarily.

File: user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.test.tsx [hunk 4]
```diff
@@ -103,20 +104,21 @@ describe('CaseNoteFormModal - Simple Tests', () => {
...
+    const contentInput = document.querySelector(CONTENT_INPUT_SELECTOR);
     const newTitle = 'New Note Title';
     const newContent = 'New Note Content';
+    const expectedContent = '<p>New Note Content</p>';
 
     await userEvent.type(titleInput, newTitle);
-    await userEvent.type(contentInput, newContent);
+    await userEvent.type(contentInput!, newContent);
 
     const submitButton = screen.getByTestId(SUBMIT_BUTTON_ID);
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.test.tsx:113]: This is a test, so we actually want it to fail if element is missing.

File: user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.test.tsx [hunk 5]
```diff
@@ -278,9 +280,9 @@ describe('CaseNoteFormModal - Simple Tests', () => {
     await userEvent.click(openButton);
 
     const titleInput = screen.getByTestId(TITLE_INPUT_ID);
-    const contentInput = screen.getByTestId(CONTENT_INPUT_ID);
+    const contentInput = document.querySelector(CONTENT_INPUT_SELECTOR);
     await userEvent.type(titleInput, 'Test Title');
-    await userEvent.type(contentInput, 'Test Content');
+    await userEvent.type(contentInput!, 'Test Content');
```
Comment (fmaddenflx) [inline user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.test.tsx:285]: Yeah maybe rather than using bang ! everywhere, we should make it a habbit to use expect() to check that something is not null or undefined.

File: user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.tsx [hunk 6]
```diff
@@ -35,10 +37,14 @@ const useThrottleCallback = (callback: () => void, delay: number) => {
...
 };
 
-export function getCaseNotesInputValue(ref: TextAreaRef | null) {
+export function getCaseNotesTitleValue(ref: InputRef | null) {
   return ref?.getValue() ?? '';
 }
 
+export function getCaseNotesContentValue(ref: RichTextEditorRef | null) {
+  return ref?.getHtml() ?? '';
+}
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/case-detail/panels/case-notes/CaseNoteFormModal.tsx:46]: We prefer self-documenting code rather than comments.  We could rename the function to something like getCaseNotesHtmlContent()

File: user-interface/src/lib/components/uswds/Input.tsx [hunk 7]
```diff
@@ -21,7 +21,7 @@ function InputComponent(props: InputProps, ref: React.Ref<InputRef>) {
   const [inputValue, setInputValue] = useState<string>(props.value || '');
   const [inputDisabled, setInputDisabled] = useState<boolean>(props.disabled ?? false);
 
-  const { includeClearButton, ariaDescription, errorMessage, ...otherProps } = props;
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/uswds/Input.tsx:24]: Thought I added an explicit use of required to the jsx for rendering the input.  If not, then we should add it.  I want it made clear that we are using required so that there is no question.

File: user-interface/src/lib/components/uswds/Input.tsx [hunk 8]
```diff
@@ -101,6 +100,7 @@ function InputComponent(props: InputProps, ref: React.Ref<InputRef>) {
         >
           <input
             {...otherProps}
+            required={required}
```
Comment (fmaddenflx) [inline user-interface/src/lib/components/uswds/Input.tsx:103]: This is where we are explicitly adding required.  So we are using the required that was destructured from props.

File: user-interface/src/lib/components/cams/RichTextEditor/SelectionService.humble.ts
Comment (fmaddenflx) [inline user-interface/src/lib/components/cams/RichTextEditor/SelectionService.humble.ts:1]: Let's add a function to this service for setting the inner html of an element with DOMPurify.

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    'onmouseover',
+    'onmousedown',
+    'onfocus',
+    'onblur',
+    'onkeydown',
+    'onkeypress',
+    'srcset',
+    'src',
+  ],
+};
```
Comment (fmaddenflx) [inline user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:26]: Let's place this config inside the selection service module with a function in the class to handle setting innerHTML.

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+      contentRef.current.innerHTML = '';
+      // Re-initialize with empty paragraph
+      editorRef.current = new Editor(contentRef.current, selectionService);
+      onChange?.('');
+    }
+  };
+
+  const getValue = () => contentRef.current?.innerText || '';
+
+  const getHtml = () => {
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:78]: I suppose perhaps we don't even need getValue at all.  Or perhaps we call getHtml getValue instead and do away with getHtml ?

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+  const setValue = (html: string) => {
+    if (contentRef.current) {
+      if (html.trim() === '') {
+        // If setting empty content, reinitialize with empty paragraph
+        contentRef.current.innerHTML = '';
+        if (editorRef.current) {
+          editorRef.current = new Editor(contentRef.current, selectionService);
+        }
+      } else {
+        contentRef.current.innerHTML = DOMPurify.sanitize(html, DOMPURIFY_CONFIG);
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:103]: why are you looping?

[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:103]: Be more DRY.

[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:103]: DRY

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            onChange?.(getHtml());
+          }}
+        ></RichTextButton>
+      </div>
+      <div
+        id={id}
+        className={`editor-content ${className || ''}`}
+        contentEditable={!inputDisabled}
+        tabIndex={0}
+        onInput={() => onChange?.(getHtml())}
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:240]: Good suggestion perhaps.

File: user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+          }}
+        ></RichTextButton>
+      </div>
+      <div
+        id={id}
+        data-testid={id}
+        className={`editor-content ${className || ''}`}
+        contentEditable={!inputDisabled}
+        tabIndex={0}
+        onInput={() => onChange?.(getHtml())}
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/RichTextEditor.tsx:240]: DRY

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.ts [hunk 14]
```diff
@@ -0,0 +1,1315 @@
...
+
+  public static cleanEmptyTags(html: string): string {
+    return html.replace(EMPTY_TAG_REGEX, '');
+  }
+
+  public static cleanHtml = (html: string) => {
+    return Editor.cleanEmptyTags(Editor.cleanZeroWidthSpaces(html));
+  };
+
+  public handleCtrlKey(e: React.KeyboardEvent<HTMLDivElement>): boolean {
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.ts:75]: While the end users currently won't be using macs, this is actually not a bad suggestion if perhaps things change, ever, and may make things easier for us developers who are all using macs.

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.ts [hunk 15]
```diff
@@ -0,0 +1,1313 @@
...
+    if (e.metaKey) {
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.ts:76]: Maybe a better approach would be to call the given Ctrl key handler for equivalent meta key combinations.

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.ts [hunk 16]
```diff
@@ -0,0 +1,1315 @@
...
+    this.selectionService.setSelectionRange(newRange);
+
+    if (parentList && [...parentList.children].every((child) => child.textContent?.trim() === '')) {
+      parentList.remove();
+    }
+
+    return true;
+  }
+
+  public handlePrintableKey(e: React.KeyboardEvent<HTMLDivElement>): boolean {
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.ts:328]: We don't need any internationalization.  This is a US only app.  Also, what is IME?

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 17]
```diff
@@ -0,0 +1,3256 @@
...
+function _insertTextAtSelection(text: string, selectionService: MockSelectionService) {
+  const selection = selectionService.getCurrentSelection();
+  const range = selection.getRangeAt(0);
+  const textNode = document.createTextNode(text);
+  range.insertNode(textNode);
+  range.setStartAfter(textNode);
+  range.collapse(true);
+  selection.removeAllRanges();
+  selection.addRange(range);
+}
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:46]: remove if not used.

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 18]
```diff
@@ -0,0 +1,3256 @@
...
+
+describe('Editor2', () => {
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:48]: yes.

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 19]
```diff
@@ -0,0 +1,3256 @@
...
+    const event = createBackspaceEvent();
+    const result = editor.handleBackspaceOnEmptyContent(event);
+
+    expect(result).toBe(true);
+    expect(event.preventDefault).toHaveBeenCalled();
+    expect(container.querySelectorAll('p')).toHaveLength(1);
+  });
+
+  test('allows deletion when multiple paragraphs exist', () => {
+    container.innerHTML = `<p>First</p><p>${ZERO_WIDTH_SPACE}</p>`;
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:844]: D.R.Y!!!!!!!!!!!!!!!!!!!!!!

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 20]
```diff
@@ -0,0 +1,3256 @@
...
+      range.setStart(span.firstChild!, 0);
+      range.collapse(true);
+      selectionService.setSelectionRange(range);
+
+      // Mock the createRange method to return a range that will trigger the non-text node branch
+      const originalCreateRange = selectionService.createRange;
+      selectionService.createRange = vi.fn().mockImplementation(() => {
+        const mockRange = originalCreateRange.call(selectionService);
+        // Force the code to go through the non-text node branch in the final paragraph creation
+        return mockRange;
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:1765]: I only need to be told once Sourcery.  Simmer down now.

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 21]
```diff
@@ -0,0 +1,3256 @@
...
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:2013]: 8-|

File: user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts [hunk 22]
```diff
@@ -0,0 +1,3256 @@
...
+
+      const event = createBackspaceEvent();
+      const result = editor.handleDeleteKeyOnList(event);
+
+      expect(result).toBe(true);
+      expect(event.preventDefault).toHaveBeenCalled();
+      // The method removes the empty list item but doesn't create a paragraph since the list still has content
+      expect(container.querySelectorAll('li')).toHaveLength(1);
+      expect(container.querySelector('li')!.textContent).toBe('First item');
+    });
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply user-interface/src/lib/components/cams/RichTextEditor/Editor.test.ts:2435]: very good idea.
